crates/custody/src/encrypted.rs

```diff
@@ -0,0 +1,264 @@
+use penumbra_proto::custody::v1::{self as pb, AuthorizeResponse};
+use rand_core::OsRng;
+use serde::{Deserialize, Serialize};
+use serde_with::{formats::Uppercase, hex::Hex};
+use tokio::sync::OnceCell;
+use tonic::{async_trait, Request, Response, Status};
+
+use crate::{soft_kms, terminal::Terminal, threshold};
+
+mod encryption {
+    use anyhow::anyhow;
+    use chacha20poly1305::{
+        aead::{AeadInPlace, NewAead},
+        ChaCha20Poly1305,
+    };
+    use rand_core::CryptoRngCore;
+
+    /// Represents a password that has been validated for length, and won't cause argon2 errors
+    #[derive(Clone, Copy)]
+    pub struct Password<'a>(&'a str);
+
+    impl<'a> Password<'a> {
+        /// Create a new password, validating its length
+        pub fn new(password: &'a str) -> anyhow::Result<Self> {
+            anyhow::ensure!(password.len() < argon2::MAX_PWD_LEN, "password too long");
+            Ok(Self(password))
+        }
+    }
+
+    impl<'a> TryFrom<&'a str> for Password<'a> {
+        type Error = anyhow::Error;
+
+        fn try_from(value: &'a str) -> Result<Self, Self::Error> {
+            Self::new(value)
+        }
+    }
+
+    // These can be recomputed from the library, at the cost of importing 25 billion traits.
+    const SALT_SIZE: usize = 32;
+    const TAG_SIZE: usize = 16;
+    const KEY_SIZE: usize = 32;
```

Comment:
The choice of Argon2id makes sense. I looked at the default parameters from the library we're using, and they are more stringent vs [RFC9106](https://datatracker.ietf.org/doc/html/rfc9106#name-parameter-choice) _except_ for the default choice of p=1 - the first recommended choice in RFC9106 suggests p=4

Reply:
We probably want to just hard-code the RFC params for two reasons:
1. Matching the RFC is nice
2. The default might change underneath us if we upgrade the argon dep, breaking people's stuff.

Reply:
yep this is a good call